### PR TITLE
Feature/openid connect

### PR DESCRIPTION
**Pull request by dylanturn**
Adds OIDC support without needing sidecars/proxies. 

Resolves: https://github.com/cdr/code-server/issues/2636

### CODE REVIEW

File: src/node/http.ts
Comment (jawnsy) [inline src/node/http.ts:75]: It looks like we might want to use `console.debug` and add some other text here? It's definitely helpful to have debug output as figuring out OIDC integration problems can be... a challenge :sweat_smile: but we'd some other text that we can search for, so we know where the logs are coming from.

maybe something like:

```javascript
console.debug("authenticated using OpenID Connect as", req.oidc.user)
```

other debug info like the returned OIDC claims seems useful too? they're secret credentials, but since code-server is meant for a single user, maybe logging those is OK?

**Review by jawnsy — COMMENTED** [review]
Hey there,

Thanks so much for your contribution! This looks really great, but I'm not one of the primary maintainers of code-server, so I'll leave the review to others :smile:

File: src/node/http.ts
Comment (dylanturn) [reply src/node/http.ts:75]: Yes, very sorry, I'll include this change along with the updates to yarn lockfile.

Comment (jawnsy) [reply src/node/http.ts:75]: No apologies needed! Thanks so much for your contribution! :heart:

**Comment by code-asher** [comment]
Thanks for the contribution! I think we'll need to discuss whether we want to support more authentication methods. The main reason we've been against it is that we felt it wasn't worth the maintenance effort when you could just use a purpose-built proxy.

For example, I personally wish we had recommended an external reverse proxy instead of building one into code-server so I'm reluctant to do a similar thing here.

But I'm definitely curious to everyone's thoughts and this seems to have been requested quite a few times so the desire certainly seems to be there. So maybe we put it to a vote or something.

@jsjoeio @oxy @jawnsy

**Comment by oxy** [comment]
Hi there! I'm not really against the PR and I think its a great addition! My only concern is a lack of documentation around how to set up OpenID authentication, what the secrets are, how it would work...

I've never touched OpenID before, and in my quick Googling it didn't sound really easy to me either - so I'd really appreciate if you could also write some documentation on how to set it up :sweat_smile:

**Comment by dylanturn** [comment]
> Hi there! I'm not really against the PR and I think its a great addition! My only concern is a lack of documentation around how to set up OpenID authentication, what the secrets are, how it would work...
> 
> I've never touched OpenID before, and in my quick Googling it didn't sound really easy to me either - so I'd really appreciate if you could also write some documentation on how to set it up 😅

Funny you should mention that, I'm busy writing documentation that explains how to setup an OpenID application on Auth0, Okta, and Keycloak. I've also been considering a small code update that would make the `openid-group-claim` argument optional, simplifying the OpenID application configuration process for those who don't require group/role based access.

**Comment by jsjoeio** [comment]
First, I just want to say thanks for the contribution! We really appreciate your input and and want to encourage more people to contribute to code-server.

Echoing @code-asher, biggest concern with something like this is:
1. What precedent does it establish for folks who want other types of authentication methods in code-server?
2. What would the maintenance for something like this look like for us a maintainers?

I wish we had some type of plugin ecosystem so that people could build _on top_ of code-server, which would be a nice alternative.

I think we'll need to discuss this as a group before we decide what to do. In the meantime, thank you for being patient!

**Comment by dylanturn** [comment]
I'm not really sure what my part in this conversation is meant to be, but for what it's worth here are my thoughts on the two concerns you raised.

> * What precedent does it establish for folks who want other types of authentication methods in code-server?
Would adding LDAP/SAML/Kerberos authentication be problematic provided the implementations were simple and leveraged well known and trusted libraries? 

> * What would the maintenance for something like this look like for us a maintainers?
Given that the implementation uses Auth0s `express-openid-connect` library and only consists of around 100 lines of actual logic, I would imagine this would be maintained like any of the other functionality that leverages 3rd party libraries.

Judging by the contents of `typings/pluginapi.d.ts` and `typings/ipc.d.ts` it looks like there has been work done to create some kind of plugin system. That said, from what I can tell the implemented plugins wouldn't be capable of authenticating users, which makes sense since allowing plugins to authenticate users could prove to be a security risk.

**Comment by dylanturn** [comment]
For what it's worth, I just merged in the following changes to this branch:
- Removed the OIDC group claims requirement to allow for simpler OpenID-Connect configurations.
- Added documentation that walks through the process of configuring OpenID-Connect authentication endpoints in Okta, Auth0, and Keycloak.

File: src/node/http.ts
Comment (oxy) [inline src/node/http.ts:76]: Thank you so much for your PR! Only nit is that we'd prefer you use `logger.debug` instead of `console.debug` for consistency with the rest of our codebase.

For example: `logger.debug("User authenticated with OpenID Connect", field("user", req.oidc.user))`

Sorry about the `console.debug` suggestion earlier :sweat_smile:

File: src/node/http.ts
Comment (oxy) [inline src/node/http.ts:90]: Same as above, `logger.debug`

File: src/node/http.ts
Comment (oxy) [inline src/node/http.ts:98]: `logger.debug`

File: src/node/http.ts
Comment (oxy) [inline src/node/http.ts:104]: `logger.debug`

**Review by oxy — CHANGES_REQUESTED** [review]
Thanks a lot for your contribution! Only nit is the `logger.debug`/`console.debug` difference; sorry about misleading you earlier :sweat_smile:

We've had this feature requested from us several times; I'll test this myself and merge this soon - I figure that since its a relatively isolated change (~150 lines of code), and a library made by a relatively trusted org (Auth0), it wouldn't cause a large maintenance burden.

File: src/node/http.ts
Comment (dylanturn) [reply src/node/http.ts:76]: haha, no worries, consistency is king! I'll get it updated and pushed at some point today.

Comment (dylanturn) [reply src/node/http.ts]: The updates have been made. Let me know if there's anything else that needs to be tweaked.

File: src/node/http.ts
Comment (dylanturn) [reply src/node/http.ts]: Done :-)

File: src/node/http.ts
Comment (dylanturn) [reply src/node/http.ts]: Done :-)

File: src/node/http.ts
Comment (dylanturn) [reply src/node/http.ts]: Done :-)

File: src/node/http.ts
Comment (dylanturn) [reply src/node/http.ts]: I pushed one last change to update the logger.debug calls to use field() instead of whatever it is I did.

**Comment by dylanturn** [comment]
> Thanks a lot for your contribution! Only nit is the `logger.debug`/`console.debug` difference; sorry about misleading you earlier 😅
> 
> We've had this feature requested from us several times; I'll test this myself and merge this soon - I figure that since its a relatively isolated change (~150 lines of code), and a library made by a relatively trusted org (Auth0), it wouldn't cause a large maintenance burden.

Is there anything you need from me before this can be merged?

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:5]:
```suggestion
More information regarding how OpenID Connect works can be found here:
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:3]:
```suggestion
The purpose of this guide is to demonstrate the process of configuring an OpenID application as an authentication and authorization source for Code-Server.
```

Comment (vapurrmaid) [inline docs/openid-auth.md:3]:
```suggestion
The purpose of this guide is to demonstrate the process of configuring an OpenID Application as an authentication and authorization source for code-server.
```

We refer to code-server in lowercase format, see: https://github.com/cdr/styleguide/blob/master/product-names.md#capitalized-terms

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:9]:
```suggestion
- <https://openid.net/connect/faq/>
- <https://auth0.com/docs/protocols/openid-connect-protocol>
- <https://developer.okta.com/blog/2019/10/21/illustrated-guide-to-oauth-and-oidc>
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:13]:
```suggestion
- Either Keycloak has been deployed or an existing Auth0/Okta account exists.
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:16]: Either `[]()` format, or add `<>` around the URLs

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:17]:
```suggestion
- A value needs to be generated for `openid-secret`. This value will be used to encrypt the OpenID Connect session token.
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:19]:
```suggestion
  - Windows: <https://activedirectoryfaq.com/2017/08/creating-individual-random-passwords/>
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:25]:
```suggestion
1. Navigate to the application management section of your Auth0 dashboard at `https://manage.auth0.com/dashboard/us/{{auth0_account_name}}/applications`
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:26]:
```suggestion
2. Click the **Create Application** button in the top right of the page.
```

See: <https://github.com/cdr/styleguide/blob/master/text-formatting.md#bold>

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:27]:
```suggestion
3. Either provide a name for this application or use the default, then select **Regular Web Application** as the application type and click the blue **Create** button.
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:34]:
```suggestion
1. Make note of the `Client ID` value. This value will be used in the code-server `openid-client-id` configuration variable.
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:1]:
```suggestion
# Configuring OpenID authentication
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:21]:
```suggestion
## Auth0 OpenID Connect application setup
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:23]:
```suggestion
### Creating the application
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:32]:
```suggestion
### Gather the client ID
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:46]:
```suggestion
### Example Auth0 code-server configuration
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:50]:
```suggestion
**NOTE:** The `openid-issuer-base-url` should be set to the value of the **Domain** field of the Okta application.
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:65]:
```suggestion
## Okta OpenID Connect application setup
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:67]:
```suggestion
### Creating the application
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:69]:
```suggestion
1. Navigate to the **Add Application** page of your Okta dashboard at `https://{{okta_account_name}}-admin.okta.com/admin/apps/add-app`
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:70]:
```suggestion
2. Click the **Create New App** button located in the upper right of the page.
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:71]:
```suggestion
3. Set the **Platform** to **Web** and **Sign on method** to **OpenID Connect**. Then click **Create**
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:76]:
```suggestion
### Update the application
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:78]:
```suggestion
1. Update the **Application name** field with the desired name for this application.
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:79]:
```suggestion
2. Update the **Login redirect URIs** and **Logout redirect URIs (Optional)** fields so that they point to the correct code-server endpoint then click **Save**
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:81]:
```suggestion
3. To update the **Allowed grant types** start by navigating to the **General Settings** section and clicking the **Edit** link in the top right corner of the section card.
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:82]:
```suggestion
4. Once in the edit view, locate the **Allowed grant types** checkbox list and make sure the boxes for **Implicit (Hybrid)** and **Allow ID Token with implicit grant type** are checked, then scroll to the bottom of the page and click **Save**.
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:84]:
```suggestion
5. Lastly, ensure a user is assigned to this application by navigating to the **Assignments** tab, clicking on **Assign**, then selecting **Assign to People**.
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:87]:
```suggestion
### Gather the client ID
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:89]:
```suggestion
1. Make note of the `Client ID` value. This value will be used in the code-server `openid-client-id` configuration variable.
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:94]:
```suggestion
### Example code-server configuration
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:98]:
```suggestion
**NOTE:** The `openid-issuer-base-url` should be set to the value of the **Okta domain** field of the Okta application.
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:113]:
```suggestion
## Keycloak OpenID Connect client setup
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:115]:
```suggestion
### Creating the client
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:124]:
```suggestion
### Update the client name
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:133]:
```suggestion
### Example code-server configuration
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:117]:
```suggestion
1. Once logged into the Keycloak instance, navigate to the **Clients** page by clicking the **Clients** link in the navigation menu on the left side of the page.
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:118]:
```suggestion
2. Begin the client creation process by clicking the **Create** button located in the top right corner of the client list.
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:119]:
```suggestion
3. Fill out the client creation fields then click **Save** - Client ID**: This is the value that will later be populated in `openid-client-id`. This value is entirely up the the user or process creating the client. - **Client Protocol**: This value should be set to **openid-connect** - **Root URL**: This field should be populated with the code-server base URL.
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:126]:
```suggestion
1. Once the **Save** button in the **Add Client** window has been clicked, the client will be created and the page will be redirected to the client settings view. From inside that view proceed to name the newly create client by populating the **Name** field.
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:127]:
```suggestion
2. Enable implicit flow by changing **Implicit Flow Enabled** from **OFF** to **ON**.
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:128]:
```suggestion
3. Scroll to the bottom of the page and click **Save**.
```

File: docs/openid-auth.md
Comment (vapurrmaid) [inline docs/openid-auth.md:139]:
```suggestion
**NOTE:** The `openid-client-id` value should be set to the value given to the **Client ID** when the client was first created.
```

**Review by vapurrmaid — COMMENTED** [review]
Hey, :clap: thanks so much for a rad contribution. I took a look at _just_ the documentation and have various _very minor nitpicks_ so that it aligns with our style guide at Coder.

Check it out here if you'd like to: <https://github.com/cdr/styleguide>

Namely, we are trying to ensure the following styles:

- code-server is always referred to in lowercase
- headings are sentence case
- UI elements use **bold**, but not italics

:clap:

**Comment by dylanturn** [comment]
> Hey, 👏 thanks so much for a rad contribution. I took a look at _just_ the documentation and have various _very minor nitpicks_ so that it aligns with our style guide at Coder.
> 
> Check it out here if you'd like to: https://github.com/cdr/styleguide
> 
> Namely, we are trying to ensure the following styles:
> 
> * code-server is always referred to in lowercase
> * headings are sentence case
> * UI elements use **bold**, but not italics
> 
> 👏

Thank you for making the changes easy to commit :-). I've applied your suggestions, let me know if there's anything else that needs to be updated.

Thank you :-)

**Comment by vapurrmaid** [comment]
> > Hey, clap thanks so much for a rad contribution. I took a look at _just_ the documentation and have various _very minor nitpicks_ so that it aligns with our style guide at Coder.
> > Check it out here if you'd like to: https://github.com/cdr/styleguide
> > Namely, we are trying to ensure the following styles:
> > 
> > * code-server is always referred to in lowercase
> > * headings are sentence case
> > * UI elements use **bold**, but not italics
> > 
> > clap
> 
> Thank you for making the changes easy to commit :-). I've applied your suggestions, let me know if there's anything else that needs to be updated.
> 
> Thank you :-)

AFAICT, things are good to go - thanks for taking so much care with this contribution!!!

**Comment by code-asher** [comment]
Hey @dylanturn thank you for all the work you've done here!

Since my first comment we've been internally continuing the discussion on whether to accept PRs that introduce alternative authentication methods and we ultimately decided we would stick with the existing policy.

Part of the reason was because we want to limit core features to ones that directly impact the development experience of code-server itself (I feel that authentication is more an infrastructure problem) and the other part was that we didn't feel prepared to own and test the feature especially as we'd need to test manually.

I'm sorry we didn't reach this decision sooner before you put all this work into addressing the comments on your PR! :sob: 

But all is not lost. I think it makes sense to adapt this code to a plugin. I think we can get there most of the way with the existing API but we will probably need to make some changes to do it properly.

I'm not sure exactly what the API will need to look like so we're very much open to ideas. The current API is not considered stable so we can change it as much as we like before finally making it public. Once that's completed we can re-open and adapt the code.

**Pull request closed without merging**
